PR: https://github.com/odoo/odoo/pull/

From: 1d777d6d952d462b13bdcd8e39836fd339d26107
From: Martin Geubelle
Date: 2016-03-10 12:41:07

Structural Changes: 1
Total Changes: 57

[IMP] payment, payment_*: acquirers installation

The installation of a new acquirer was a bit complicated : from settings,
check the acquirer, then apply (install the module) then list view of
acquirer and finally edit it in form view.

This needed to be simplified. The payment acquirers are pre-filled
in payment. From the kanban view an `Install` button installs and
redirects to the form field.

================================= pseudo patch: =================================

--- a/addons/payment/models/__init__.py
+++ b/addons/payment/models/__init__.py
@@ -1,5 +1,4 @@
 # -*- coding: utf-8 -*-
 
 import payment_acquirer
-import res_config
 import res_partner

--- a/addons/payment/models/payment_acquirer.py
+++ b/addons/payment/models/payment_acquirer.py
@@ -54,14 +54,14 @@ class PaymentAcquirer(osv.Model):
     _order = 'sequence'
 
     def _get_providers(self, cr, uid, context=None):
-        return []
+        return [('manual', 'Manual Configuration')]
 
     # indirection to ease inheritance
     _provider_selection = lambda self, *args, **kwargs: self._get_providers(*args, **kwargs)
 
     _columns = {
         'name': fields.char('Name', required=True, translate=True),
-        'provider': fields.selection(_provider_selection, string='Provider', required=True),
+        'provider': fields.selection(_provider_selection, string='Provider', required=True, default='manual'),
         'company_id': fields.many2one('res.company', 'Company', required=True),
         'pre_msg': fields.html('Help Message', translate=True,
                                help='Message displayed to explain and help the payment process.'),
@@ -92,6 +92,9 @@ class PaymentAcquirer(osv.Model):
         'fees_int_fixed': fields.float('Fixed international fees'),
         'fees_int_var': fields.float('Variable international fees (in percents)'),
         'sequence': fields.integer('Sequence', help="Determine the display order"),
+        'module_id': fields.many2one('ir.module.module', string='Corresponding Module'),
+        'module_state': fields.related('module_id', 'state', type='char', string='Installation State'),
+        'description': fields.html('Description'),
     }
 
     image = openerp.fields.Binary("Image", attachment=True,
@@ -315,6 +318,20 @@ class PaymentAcquirer(osv.Model):
         self.write(cr, uid, prod_ids, {'environment': 'test'}, context=context)
         self.write(cr, uid, test_ids, {'environment': 'prod'}, context=context)
 
+    def button_immediate_install(self, cr, uid, ids, context=None):
+        acquirer_id = self.browse(cr, uid, ids, context=context)
+        if acquirer_id.module_id and acquirer_id.module_state != 'installed':
+            acquirer_id.module_id.button_immediate_install()
+            context['active_id'] = ids[0]
+            return {
+                'view_type': 'form',
+                'view_mode': 'form',
+                'res_model': 'payment.acquirer',
+                'type': 'ir.actions.act_window',
+                'res_id': ids[0],
+                'context': context,
+            }
+
 
 class PaymentTransaction(osv.Model):
     """ Transaction Model. Each specific acquirer can extend the model by adding

--- a/addons/payment/models/res_config.py
+++ b/None
@@ -1,35 +0,0 @@
-# -*- coding: utf-8 -*-
-
-from openerp.osv import fields, osv
-
-
-class AccountPaymentConfig(osv.TransientModel):
-    _inherit = 'account.config.settings'
-
-    _columns = {
-        'module_payment_transfer': fields.boolean(
-            'Wire Transfer',
-            help='-This installs the module payment_transfer.'),
-        'module_payment_paypal': fields.boolean(
-            'Paypal',
-            help='-This installs the module payment_paypal.'),
-        'module_payment_ogone': fields.boolean(
-            'Ogone',
-            help='-This installs the module payment_ogone.'),
-        'module_payment_adyen': fields.boolean(
-            'Adyen',
-            help='-This installs the module payment_adyen.'),
-        'module_payment_buckaroo': fields.boolean(
-            'Buckaroo',
-            help='-This installs the module payment_buckaroo.'),
-        'module_payment_authorize': fields.boolean(
-            'Authorize.Net',
-            help='-This installs the module payment_authorize.'),
-        'module_payment_sips': fields.boolean(
-            'Sips',
-            help='-This installs the module payment_sips.'),
-    }
-
-    _defaults = {
-        'module_payment_transfer': True
-    }
